src/Microsoft.PowerShell.Commands.Management/commands/management/Process.cs

```diff
@@ -736,53 +737,89 @@ private static string RetrieveProcessUserName(Process process)
+                if (!Interop.Windows.OpenProcessToken(process.Handle, Interop.Windows.TOKEN_QUERY, out processTokenHandler))
                 {
-                    error = Marshal.GetLastWin32Error();
-                    if (error == Win32Native.ERROR_INSUFFICIENT_BUFFER)
-                    {
-                        Marshal.FreeHGlobal(tokenUserInfo);
-                        tokenUserInfo = Marshal.AllocHGlobal(tokenInfoLength);
-
-                        if (!Win32Native.GetTokenInformation(processTokenHandler, Win32Native.TOKEN_INFORMATION_CLASS.TokenUser, tokenUserInfo, tokenInfoLength, out tokenInfoLength)) { return null; }
-                    }
-                    else
-                    {
-                        return null;
-                    }
+                    return null;
                 }
 
-                var tokenUser = Marshal.PtrToStructure<Win32Native.TOKEN_USER>(tokenUserInfo);
+                // Set the default length to be 256, so it will be sufficient for most cases.
+                const int StartLength =
+#if DEBUG
+                    // In debug, validate ArrayPool growth.
+                    1;
+#else
+                    256;
+#endif
+                Span<byte> tokenData = stackalloc byte[StartLength];
+                Interop.Windows.TOKEN_USER tokenUser;
+                byte[] rentedArray = null;
 
                 // Max username is defined as UNLEN = 256 in lmcons.h
                 // Max domainname is defined as DNLEN = CNLEN = 15 in lmcons.h
                 // The buffer length must be +1, last position is for a null string terminator.
-                int userNameLength = 257;
-                int domainNameLength = 16;
-                Span<char> userNameStr = stackalloc char[userNameLength];
-                Span<char> domainNameStr = stackalloc char[domainNameLength];
-                Win32Native.SID_NAME_USE accountType;
+                const int UserNameLength = 257;
+                const int DomainNameLength = 16;
+                int userNameLength = UserNameLength;
+                int domainNameLength = DomainNameLength;
+                Span<char> userNameStr = stackalloc char[UserNameLength];
+                Span<char> domainNameStr = stackalloc char[DomainNameLength];
 
-                // userNameLength and domainNameLength will be set to actual lengths.
-                if (!Win32Native.LookupAccountSid(null, tokenUser.User.Sid, userNameStr, ref userNameLength, domainNameStr, ref domainNameLength, out accountType))
+                try
                 {
-                    return null;
-                }
+                    while (true)
+                    {
+                        int tokenInfoLength = tokenData.Length;
+                        unsafe
+                        {
+                            fixed (byte* pinnedTokenData = &MemoryMarshal.GetReference(tokenData))
```

Comment:
Do we need MemoryMarshal here? In the past I've just done `fixed (byte* dataPtr = data)` where `data` is a `Span<byte>`. Just wondering what benefits we get when using `MemoryMarshal.GetReference`.

Reply:
Yeah, I would usually expect to to see the pattern-based fixed statement instead, which gets a reference by calling `span. GetPinnableReference()`.

Reply:
`MemoryMarshal.GetReference` does avoid a check for an empty span, but in terms in performance this is not measurable. See https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Span.cs,7ff7986fc34855d6

Reply:
We need `pinnedTokenData` as raw buffer to do in-place cast to `tokenUser` and send it (the same buffer!) without additional pinning to next pinvoke.

Reply:
@jborean93 is saying we can use the pattern-based fixed statement, i.e. `fixed (byte* pinnedTokenData = tokenData)`.

Reply:
There can be second iteration with rented array. Will this work correctly too?